Convert Orion and Hera modulefiles to LUA format and hpc-stack

KateFriedman-NOAA opened the pull request:
**Description**

This PR converts the Orion and Hera modulefiles to LUA format (rename to add `.lua`) and to use hpc-stack. This PR also includes the following:

- initial build updates to get workflow-owned codes to build with hpc-stack on Orion/Hera (more of this to come, waiting for some library module versions to be installed)
- create new `$target.ver` files for Orion and Hera; these new versions files will set hpc-stack modules and can be used to set other library module versions that may be different on specific platforms (due to need)
- cleanup the Hera, Orion, and Jet blocks in machine-setup.sh
- delete unneeded OznMonBuild and RadMonBuild modulefiles for Orion and WCOSS-Dell (didn't exist for Hera or WCOSS2)

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

By running individual build scripts for codes that aren't awaiting a library version to be installed in hpc-stack or have a viable alternative to use for now. Can successfully build the following execs on Orion/Hera with these updates:

- enkf_chgres_recenter.x
- fbwndgfs
- gaussian_sfcanl.exe
- gfs_bufr
- regrid_nemsio
- tocsbufr

Confirmed updates do not impact building these codes on WCOSS2 by running the builds on WCOSS2 in a clone of my fork copy of `dev_v16`.
  
**Checklist**

- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings

KateFriedman-NOAA committed:
Rename Orion modulefiles for LUA format

- add the ".lua" extension to the Orion modulefiles

Refs: #639

KateFriedman-NOAA committed:
Remove uneeded modulefiles for Orion/WCOSS-Dell

- cleaned out modulefiles that are no longer needed for OznMon/Radmon
- remove errant workflow_utils.orion.lua (wrong branch)

Refs: #639

KateFriedman-NOAA committed:
Update Orion modulefiles to LUA format and hpc-stack

Refs: #639

KateFriedman-NOAA committed:
Add myFC variable to WCOSS2 modulefiles

- add "setenv("myFC","ftn")" to the modulefiles for fbwndgfs
and storm_reloc on WCOSS2
- needed to set FC on WCOSS2 now that other platforms have
to set FC differently (e.g. mpiifort on Orion)

Refs: #639

KateFriedman-NOAA committed:
Update fbwndgfs and storm_reloc builds to use myFC

- update fbwndgfs makefile.GENERIC to use myFC variable
- update build_tropcy_NEMS.sh to use myFC variable
- also update build_tropcy_NEMS.sh to adjust JASPER_LIB variable
and remove wcoss2 check for setting SIGIO_LIB4 and SIGIO_INC4

Refs: #639

KateFriedman-NOAA committed:
Cleanup machine-setup.sh

- cleanup Orion, Hera, and Jet blocks to just set target and
do a module purge
- add $target.ver sourcing after build.ver sourcing

Refs: #639

KateFriedman-NOAA committed:
Update gfs_bufrsnd build for hpc-stack

- update the gfs_bufr build and makefile to use hpc-stack library variables

Refs: #639

KateFriedman-NOAA committed:
Update machine-setup.sh target version sourcing

- wrap $target.ver sourcing in block that checks if user is not on
WCOSS2 and source if so
- do not need or have wcoss2.ver file at this time

Refs: #639

KateFriedman-NOAA committed:
Update NETCDF_LDFLAGS_F in gaussian_sfcanl build

- adjust NETCDF_LDFLAGS_F flags in gaussian_sfcanl.fd/makefile.sh
to flip -lnetcdf/-lnetcdff order, add -lhdf5_hl flag, and add -lz flag

Refs: #639

KateFriedman-NOAA committed:
Add $target.ver sourcing for Orion/Hera to load_fv3gfs_modules.sh

- need to now source the target-specific version files before
loading modulebase.$target.lua on Hera and Orion

Refs: #639

KateFriedman-NOAA committed:
Update build_regrid_nemsio.sh with NETCDF_LDFLAGS

- add NETCDF_LDFLAGS flags to successfully build regrid_nemsio

Refs: #639

KateFriedman-NOAA committed:
Rename Hera modulefiles to add .lua extension

Refs: #639

KateFriedman-NOAA committed:
Create new $target.ver files for Hera and Orion

- under the versions folder create new hera.ver and orion.ver
- new target-specific version files will be used to set hpc-stack
modules and versions, as well potentially override other versions
as needed on the specific platforms

Refs: #639

KateFriedman-NOAA committed:
Remove unneeded modulefile.grib_util.hera.lua

Refs: #639

KateFriedman-NOAA committed:
Convert Hera modulefiles to LUA format and hpc-stack

- convert the contents of the newly renamed Hera modules to use LUA format
- convert Hera modulefiles to load hpc-stack modules

Refs: #639

WalterKolczynski-NOAA commented on modulefiles/fv3gfs/enkf_chgres_recenter.hera.lua:
```diff
@@ -0,0 +1,20 @@
+load(pathJoin("hpc-intel", os.getenv("hpc_intel_ver")))
+load(pathJoin("hpc-impi", os.getenv("hpc_impi_ver")))
+
+load(pathJoin("hdf5", os.getenv("hdf5_ver")))
```
TCL version didn't load the NetCDF module. Necessary?

KateFriedman-NOAA replied:
Don't need to load HDF5 nor netcdf for this build. Removed both from modulefiles for Orion/WCOSS2 and reran builds to confirm. Will remove those two module loads from all enkf_chgres_recenter.*.lua modulefiles. Commited to my fork branch @ 28d1a7b.

WalterKolczynski-NOAA commented on modulefiles/gfs_fbwndgfs.hera.lua:
```diff
@@ -0,0 +1,19 @@
+
+load(pathJoin("bacio", os.getenv("bacio_ver")))
+load(pathJoin("w3nco", os.getenv("w3nco_ver")))
+load(pathJoin("nemsio", os.getenv("nemsio_ver")))
```
nemsio and sigio weren't loaded in the TCL version

KateFriedman-NOAA replied:
Removed them in my fork branch @ 75fb6da

WalterKolczynski-NOAA commented on modulefiles/gfs_fbwndgfs.orion.lua:
```diff
@@ -0,0 +1,19 @@
+
+load(pathJoin("bacio", os.getenv("bacio_ver")))
+load(pathJoin("w3nco", os.getenv("w3nco_ver")))
+load(pathJoin("nemsio", os.getenv("nemsio_ver")))
```
nemsio and sigio weren't loaded in the TCL verison

KateFriedman-NOAA replied:
Removed them in my fork branch @ 75fb6da

WalterKolczynski-NOAA commented on modulefiles/module_base.hera.lua:
```diff
@@ -0,0 +1,35 @@
+load(pathJoin("hpc", os.getenv("hpc_ver")))
+load(pathJoin("hpc-intel", os.getenv("hpc_intel_ver")))
+load(pathJoin("hpc-impi", os.getenv("hpc_impi_ver")))
+
```
TCL version was loading hpss module

KateFriedman-NOAA replied:
Thanks for catching this! This is what I get for using the Orion one as the base for the new Hera one (Orion didn't have/need a HPSS module). Have added the `hpss` module into `module_base.hera.lua` and added `hpss_ver` to `hera.ver` @ 7425fd6

WalterKolczynski-NOAA commented on modulefiles/module_base.hera.lua:
```diff
@@ -0,0 +1,35 @@
+
+load(pathJoin("python", os.getenv("python_ver")))
+load(pathJoin("gempak", os.getenv("gempak_ver")))
+load(pathJoin("perl", os.getenv("perl_ver")))
```
perl, libjpeg, udunits, bufr_dump, and util_shared weren't loaded in TCL version

KateFriedman-NOAA replied:
The `perl` module is for the GSI monitoring package.

The `libjpeg` module is a prereq for the `grib_util` module in the WCOSS2 stack. It's not listed as a prereq for the same `grib_util` module on either Hera or Orion. Asked Kyle to install `libjpeg` in the [same issue comment](https://github.com/NOAA-EMC/hpc-stack/issues/379#issuecomment-1033829965) as `udunits`.

The `udunits` is needed for something in the system on WCOSS2, although I'm having trouble finding the reason in my email. Hera has the same version as WCOSS2 and [I've already asked Kyle to add the same version to the stack on Orion](https://github.com/NOAA-EMC/hpc-stack/issues/379#issuecomment-1033829965).

The `bufr_dump` module is needed by the WAFS gcip job (see: https://github.com/NOAA-EMC/global-workflow/issues/399#issuecomment-923103405). Will need this if supporting WAFS off WCOSS2.

The `util_shared` module is needed on WCOSS2 for the wave bulletins job. See this comment: https://github.com/NOAA-EMC/global-workflow/issues/399#issuecomment-923103405. Will need this installed on Hera/Orion too if we plan to support the wave bulletins job off WCOSS2.

KateFriedman-NOAA replied:
Added `bufr_dump` and `util_shared` to my open hpc-stack issue: https://github.com/NOAA-EMC/hpc-stack/issues/379#issuecomment-1036361587

WalterKolczynski-NOAA commented on modulefiles/module_base.orion.lua:
```diff
@@ -0,0 +1,35 @@
+
+load(pathJoin("python", os.getenv("python_ver")))
+load(pathJoin("gempak", os.getenv("gempak_ver")))
+load(pathJoin("perl", os.getenv("perl_ver")))
```
perl, libjpeg, udunits, bufr_dump, and util_shared weren't loaded in TCL version

KateFriedman-NOAA replied:
See reply to above comment about `module_base.hera.lua`.

WalterKolczynski-NOAA commented on sorc/build_regrid_nemsio.sh:
```diff
@@ -18,12 +18,13 @@ export F90=${FCMP}
 export DEBUG="-traceback -g" #-O0 #-C #-fp-stack-check #-check all -fp-stack-check
 
-if [ $target != hera ]; then
+if [ $target = wcoss2 ]; then
```
If we're going to use this, why aren't we just using it on every machine?

KateFriedman-NOAA replied:
So this:
```
  LIBnetcdf=`$NETCDF/bin/nf-config --flibs`
  INCnetcdf=`$NETCDF/bin/nf-config --fflags`
  export NETCDF_LDFLAGS=$LIBnetcdf
  export NETCDF_INCLUDE=$INCnetcdf
```
...does not work on Orion, I get netcdf errors when building. Forgot to dig into this further before committing. 

Top few lines of error:
```
/apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/netcdf/4.7.4/lib/libnetcdff.a(nf_attio.o): In function `nf_put_att_text_':
nf_attio.F90:(.text+0xae): undefined reference to `nc_put_att_text'
/apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/netcdf/4.7.4/lib/libnetcdff.a(nf_attio.o): In function `nf_put_att_text_a_':
nf_attio.F90:(.text+0x17e): undefined reference to `nc_put_att_text'
/apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/netcdf/4.7.4/lib/libnetcdff.a(nf_attio.o): In function `nf_put_att_int1_':
nf_attio.F90:(.text+0x254): undefined reference to `nc_put_att_schar'
```

By using this on Orion:

```
export NETCDF_LDFLAGS="-L${NETCDF}/lib -lnetcdff -lnetcdf -L${HDF5_LIBRARIES} -lhdf5_hl -lhdf5 -lz"
```
...the final compilation command is good:
```
mpiifort  kinds.o constants.o physcons.o mpi_interface.o namelist_def.o variable_interface.o netcdfio_interface.o interpolation_interface.o gfs_nems_interface.o fv3_interface.o   -I/apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/nemsio/2.5.2/include main.f90 /apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/nemsio/2.5.2/lib/libnemsio.a /apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/bacio/2.4.1/lib/libbacio_4.a /apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/w3nco/2.4.1/lib/libw3nco_d.a /apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/sp/2.3.3/lib/libsp_4.a -L/apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/netcdf/4.7.4/lib -lnetcdff -lnetcdf -L/apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/hdf5/1.10.6/lib -lhdf5_hl -lhdf5 -lz -o regrid_nemsio > log.make.regrid_nemsio
```
If I try to use this on Orion:
```
  LIBnetcdf=`$NETCDF/bin/nf-config --flibs`
  INCnetcdf=`$NETCDF/bin/nf-config --fflags`
  export NETCDF_LDFLAGS=$LIBnetcdf
  export NETCDF_INCLUDE=$INCnetcdf
```
...the final compilation command is:
```
mpiifort  kinds.o constants.o physcons.o mpi_interface.o namelist_def.o variable_interface.o netcdfio_interface.o interpolation_interface.o gfs_nems_interface.o fv3_interface.o  -I/apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/netcdf/4.7.4/include -I/apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/netcdf/4.7.4/include -I/apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/nemsio/2.5.2/include main.f90 /apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/nemsio/2.5.2/lib/libnemsio.a /apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/bacio/2.4.1/lib/libbacio_4.a /apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/w3nco/2.4.1/lib/libw3nco_d.a /apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/sp/2.3.3/lib/libsp_4.a -L/apps/contrib/NCEP/libs/hpc-stack/intel-2018.4/impi-2018.4/netcdf/4.7.4/lib -lnetcdff -o regrid_nemsio > log.make.regrid_nemsio
```
A number of paths and flags are missing when I try to use `$NETCDF/bin/nf-config` on Orion. This works on WCOSS2.

WalterKolczynski-NOAA replied:
Okay, leave it like this for now. I don't know much about this tool, so probably something for @aerorahul to look at and see if there is something the stack team needs to fix.

KateFriedman-NOAA replied:
Note, it also doesn't work on Hera, same error. Here is the compilation command when I try:
```
mpiifort  kinds.o constants.o physcons.o mpi_interface.o namelist_def.o variable_interface.o netcdfio_interface.o interpolation_interface.o gfs_nems_interface.o fv3_interface.o  -I/scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack//intel-18.0.5.274/impi-2018.0.4/netcdf/4.7.4/include -I/scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack//intel-18.0.5.274/impi-2018.0.4/netcdf/4.7.4/include -I/scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/intel-18.0.5.274/impi-2018.0.4/nemsio/2.5.2/include main.f90 /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/intel-18.0.5.274/impi-2018.0.4/nemsio/2.5.2/lib/libnemsio.a /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/intel-18.0.5.274/bacio/2.4.1/lib/libbacio_4.a /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/intel-18.0.5.274/w3nco/2.4.1/lib/libw3nco_d.a /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/intel-18.0.5.274/sp/2.3.3/lib/libsp_4.a -L/scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack//intel-18.0.5.274/impi-2018.0.4/netcdf/4.7.4/lib -lnetcdff -o regrid_nemsio > log.make.regrid_nemsio
```

WalterKolczynski-NOAA commented on sorc/build_tropcy_NEMS.sh:
```diff
@@ -24,13 +24,11 @@ fi
-  SIGIO_INC4=$SIGIO_INC
-fi
+SIGIO_LIB4=$SIGIO_LIB
+SIGIO_INC4=$SIGIO_INC
```
Why not just change the LIBS variable to use the original variable names?

KateFriedman-NOAA replied:
Done @ 40253b9. Note this build is waiting for the `libpng`/`png` module to be updated, see https://github.com/NOAA-EMC/hpc-stack/issues/379#issuecomment-1034187891

WalterKolczynski-NOAA commented on sorc/machine-setup.sh:
```diff
@@ -210,6 +188,9 @@ fi
 # Source versions file for build
 
 . ${HOMEgfs}/versions/build.ver
+if [[ $target != wcoss2 ]] ; then
```
I'd prefer to have a blank wcoss2.ver file and not have this if-block

KateFriedman-NOAA replied:
Created `wcoss2.ver` and updated `machine-setup.sh` @ 1901756

WalterKolczynski-NOAA commented on ush/load_fv3gfs_modules.sh:
```diff
@@ -22,9 +22,11 @@ if [[ -d /lfs3 ]] ; then
 	module load module_base.jet 
 elif [[ -d /scratch1 ]] ; then
     # We are on NOAA Hera
+        source "$HOMEgfs/versions/hera.ver"
```
See above, move this outside the elif-block

KateFriedman-NOAA replied:
A `$target` variable isn't know at this point (this runs first in a job) so I don't have the ability to set:
```
source "$HOMEgfs/versions/${target}.ver"
```
I could set target in each of the if-blocks and then source `$target.ver` afterwards in the script. Thoughts?

WalterKolczynski-NOAA replied:
Let's leave it like this for here, but reevaluate whether we need both machine-setup and this script when we repeat this in develop.

KateFriedman-NOAA replied:
Agreed, want to do this and other aspects of this update better in develop when we have time to ponder this more.

WalterKolczynski-NOAA reviewed (changes requested):
Other than creating a blank wcoss2.ver file, these are mostly noting where the module list has been changed from the previous version. These changes may have been necessary due to the different lib installation (particularly if it is static vs dynamic).

Also, a few modulefiles were deleted without replacements. I'm assuming those just aren't needed.

WalterKolczynski-NOAA commented:
Also, forgot to note: some concern about the order of the loads. Though I think the new files are mostly in the order we do things now in develop, so are more likely to be correct.

KateFriedman-NOAA committed:
Remove unneeded module loads from enkf_chgres_recenter build

- remove hdf5 and netcdf module loads from enkf_chgres_recenter modulefiles
- reran builds to confirm

Refs: #639

KateFriedman-NOAA committed:
Remove unneeded modules from fbwndgfs build

- remove nemsio and sigio modules from gfs_fbwndgfs build
- reran builds to confirm

Refs: #639

KateFriedman-NOAA committed:
Add HPSS module load back into Hera module_base

- was erroneously removed, added back into module_base.hera.lua
- added companion hpss_ver to hera.ver

Refs: #639

KateFriedman-NOAA committed:
Create wcoss2.ver

- create empty wcoss2.ver to go alongside other new machine-specific
version files
- update machine-setup.sh to load $target.ver on all machines now

Refs: #639

WalterKolczynski-NOAA commented:
Can you setup tests on Hera and Orion to run over the weekend to make sure everything works?

KateFriedman-NOAA commented:
> Can you setup tests on Hera and Orion to run over the weekend to make sure everything works?

I will plan to run a test on Monday (should be able to run a fcst job at least, don't have updated components from CMs yet). Will work on test Monday after I get butterfly tests on WCOSS-Dell and WCOSS2 going (higher priority than this). Don't have time left this week to prep anything to run over weekend.

KateFriedman-NOAA committed:
Update SIGIO library variables in build_tropcy_NEMS.sh

- replace SIGIO_LIB4 with SIGIO_LIB
- replace SIGIO_INC4 with SIGIO_INC
- remove two lines that set "4" variables to non-"4" variables

Refs: #639

KateFriedman-NOAA commented:
@WalterKolczynski-NOAA I have addressed your comments, thanks for them! I will plan to test these updates at runtime on Monday and report back. Have already tested build scripts for codes that aren't waiting for stack updates or didn't have a viable alternative module version while waiting for needed version.

KateFriedman-NOAA committed:
Add prod_util_ver to orion.ver

- only available prod_util on Orion is prod_util/1.2.2
- add prod_util_ver=1.2.2 to orion.ver to override WCOSS2 version

Refs: #639

KateFriedman-NOAA committed:
Add prod_util_ver and cmake_ver to hera.ver

- add overrides for prod_util and cmake versions to hera.ver
- use available hpc-stack versions to override WCOSS2 defaults

Refs: #639

KateFriedman-NOAA committed:
Add cmake_ver to orion.ver

- override WCOSS2 default for cmake_ver and use 3.22.1 on Orion

Refs: #639
